server/grpc_bytestream.go

```diff
@@ -0,0 +1,361 @@
+		s.accessLogger.Printf(msg)
+		return status.Error(codes.Unknown, msg)
+	}
+	if rdr == nil {
+		msg := fmt.Sprintf("GRPC BYTESTREAM READ BLOB NOT FOUND: %s", hash)
+		s.accessLogger.Printf(msg)
+		return status.Error(codes.NotFound, msg)
+	}
+
+	if sizeBytes != size {
+		msg := fmt.Sprintf("Retrieved item had size %d expected %d",
+			sizeBytes, size)
+		s.accessLogger.Printf("GRPC BYTESTREAM READ FAILED (BAD ITEM?): %v", err)
+		return status.Error(codes.DataLoss, msg)
+	}
+
+	bufSize := size
+	if bufSize > maxChunkSize {
+		bufSize = maxChunkSize
+	}
+
+	buf := make([]byte, bufSize)
+
+	if req.ReadOffset > 0 {
+		seekReaderTo(rdr, req.ReadOffset)
+	}
+
+	var chunkResp bytestream.ReadResponse
+	for {
+		n, err := rdr.Read(buf)
+		if err == io.EOF {
+			s.accessLogger.Printf("GRPC BYTESTREAM READ COMPLETED %s",
+				req.ResourceName)
+			return nil
+		}
+
+		if err != nil {
+			msg := fmt.Sprintf("GRPC BYTESTREAM READ FAILED: %v", err)
+			s.accessLogger.Printf(msg)
+			return status.Error(codes.Unknown, msg)
+		}
+
+		chunkResp.Data = buf[:n]
+		err = resp.Send(&chunkResp)
+		if err != nil {
+			s.accessLogger.Printf("GRPC BYTESTREAM READ FAILED TO SEND RESPONSE: %s", err)
+			return status.Error(codes.Unknown, err.Error())
+		}
+	}
+}
+
+// Seek to offset in Reader r, from the current position.
+func seekReaderTo(r io.Reader, offset int64) {
+	switch s := r.(type) {
+	case io.Seeker:
+		s.Seek(offset, io.SeekCurrent)
+	default:
+		io.CopyN(ioutil.Discard, r, offset)
+	}
+}
+
+// Seek to offset in Writer w, from the current position.
+func seekWriterTo(w io.Writer, offset int64) {
+	switch s := w.(type) {
+	case io.Seeker:
+		s.Seek(offset, io.SeekCurrent)
+		return
+	}
+
+	blockSize := int64(len(zeros))
+	if blockSize > offset {
+		blockSize = offset
+	}
+
+	processed := int64(0)
+	for processed < offset {
+		if processed+blockSize > offset {
+			blockSize = offset - processed
+		}
+		w.Write(zeros[:blockSize])
+		processed += blockSize
+	}
+	// check: processed == offset
+}
+
+// Parse a WriteRequest.ResourceName, return the hash, size and an error.
+func (s *grpcServer) parseWriteResource(r string) (string, int64, error) {
+	// req.ResourceName is of the form:
+	// [{instance_name}/]uploads/{uuid}/blobs/{hash}/{size}[/{optionalmetadata}]
+
+	fields := strings.Split(r, "/")
+	var rem []string
+	for i := range fields {
+		if fields[i] == "uploads" {
+			rem = fields[i+1:]
+			break
+		}
+	}
+
+	if len(rem) < 4 || rem[1] != "blobs" {
+		return "", 0, status.Errorf(codes.InvalidArgument, "Unable to parse resource name: %s", r)
+	}
+
+	hash := rem[2]
+	err := s.validateHash(hash, "GRPC BYTESTREAM READ FAILED")
+	if err != nil {
+		return "", 0, err
+	}
+
+	size, err := strconv.ParseInt(rem[3], 10, 64)
+	if err != nil {
+		return "", 0, status.Errorf(codes.InvalidArgument, "Unable to parse size: %s", rem[3])
+	}
+
+	return hash, size, nil
+}
+
+func (s *grpcServer) Write(srv bytestream.ByteStream_WriteServer) error {
+
+	req, err := srv.Recv()
```

Comment:
I used two Recv calls because there's some code that should only run after the first call. I could use a single Recv call in a for loop with an if block for the first iteration, if you prefer.

Reply:
> I could use a single Recv call in a for loop with an if block for the first iteration, if you prefer.

That would be great. I think this would help readability a lot! Thanks!

Reply:
Pushed a fixup with this change.